src/activity/activity.module.ts

```diff
@@ -78,52 +108,117 @@ export class ActivityModuleImpl implements ActivityModule {
+    activity: Activity,
+    batchId: string,
+    commandIndex?: number | undefined,
+    timeout?: number | undefined,
+  ): Promise<Result[]> {
+    this.logger.info("Fetching batch results", { activityId: activity.id, batchId });
+    try {
+      const results = await this.activityApi.getExecBatchResults(activity, batchId, commandIndex, timeout);
+      this.events.emit("batchResultsReceived", activity, batchId, results);
+      return results;
+    } catch (error) {
+      this.events.emit("errorGettingBatchResults", activity, batchId, error);
+      throw error;
+    }
+  }
+  observeStreamingBatchEvents(
+    activity: Activity,
+    batchId: string,
+    commandIndex?: number | undefined,
+  ): Observable<StreamingBatchEvent> {
+    this.logger.info("Observing streaming batch events", { activityId: activity.id, batchId });
+    return this.activityApi.getExecBatchEvents(activity, batchId, commandIndex).pipe(
+      tap((event) => {
+        this.events.emit("batchEventsReceived", activity, batchId, event);
+      }),
+      catchError((error) => {
+        this.events.emit("errorGettingBatchEvents", activity, batchId, error);
+        throw error;
+      }),
+    );
+  }
 
-    this.logger.info("Created activity", {
-      activityId: activity.id,
+  async createActivity(agreement: Agreement): Promise<Activity> {
+    this.logger.info("Creating activity", {
       agreementId: agreement.id,
       provider: agreement.getProviderInfo(),
     });
-
-    return activity;
+    try {
+      const activity = await this.activityApi.createActivity(agreement);
+      this.events.emit("activityCreated", activity);
+      return activity;
+    } catch (error) {
+      this.events.emit("errorCreatingActivity", error);
+      throw error;
+    }
   }
 
-  async destroyActivity(activity: Activity, reason: string): Promise<Activity> {
-    const updated = await this.activityApi.destroyActivity(activity);
-
-    this.logger.info("Destroyed activity", {
-      activityId: updated.id,
-      agreementId: updated.agreement.id,
-      provider: updated.agreement.getProviderInfo(),
+  async destroyActivity(activity: Activity): Promise<Activity> {
+    this.logger.info("Destroying activity", {
+      activityId: activity.id,
+      agreementId: activity.agreement.id,
+      provider: activity.agreement.getProviderInfo(),
     });
+    try {
+      const updated = await this.activityApi.destroyActivity(activity);
+      this.events.emit("activityDestroyed", updated);
+      return updated;
+    } catch (error) {
+      this.events.emit("errorDestroyingActivity", activity, error);
+      throw error;
+    }
+  }
 
-    return updated;
+  async fetchActivity(activityId: Activity["id"]): Promise<Activity> {
+    this.logger.info("Fetching activity state", { activityId });
+    try {
+      const upToDateActivity = await this.activityApi.getActivity(activityId);
+      this.events.emit("activityStateChanged", upToDateActivity, upToDateActivity.getState());
```

Comment:
activity repo doesn't keep track of known activities because we can get them from yagna :/ also I don't want to jam an event emitter there. I ended up splitting the `fetchActivity` method into two - refresh and find - and emitting the event in `refresh` when the state changes. Let me know if this makes sense or we can have a call tomorrow